fv/wireguard_test.go

```diff
@@ -0,0 +1,455 @@
+		routeEntries[0], routeEntries[1] = routeEntries[1], routeEntries[0]
+	})
+
+	AfterEach(func() {
+		if CurrentGinkgoTestDescription().Failed {
+			for _, felix := range felixes {
+				felix.Exec("ip", "link", "show")
+				felix.Exec("ip", "rule", "list")
+				felix.Exec("ip", "route", "show", "table", "all")
+				felix.Exec("ip", "route", "show", "cached")
+				felix.Exec("wg")
+			}
+		}
+
+		for _, wl := range w {
+			wl.Stop()
+		}
+		for _, hwl := range hostW {
+			hwl.Stop()
+		}
+		for _, t := range tcpdumps {
+			t.Stop()
+		}
+		for _, felix := range felixes {
+			felix.Stop()
+		}
+
+		if CurrentGinkgoTestDescription().Failed {
+			infra.DumpErrorData()
+		}
+		infra.Stop()
+	})
+
+	tcpdumpMatchCount := func(felix int, matcherName string) func() int {
+		return func() int {
+			return tcpdumps[felix].MatchCount(matcherName)
+		}
+	}
+
+	Context("with Wireguard enabled", func() {
+		BeforeEach(func() {
+			for i, felix := range felixes {
+				// Check the Wireguard device exists.
+				Eventually(func() error {
+					out, err := felix.ExecOutput("ip", "link", "show", wireguardInterfaceNameDefault)
+					if err != nil {
+						return err
+					}
+					if strings.Contains(out, wireguardInterfaceNameDefault) {
+						return nil
+					}
+					return fmt.Errorf("felix %d has no Wireguard device", i)
+				}, "10s", "100ms").ShouldNot(HaveOccurred())
+			}
+		})
+
+		It("the Wireguard device should be configurable", func() {
+			enableDisableWireguard(client, false)
+
+			// wait for the Wireguard interface to disappear
+			for _, felix := range felixes {
+				Eventually(func() string {
+					out, _ := felix.ExecOutput("ip", "-d", "link", "show", wireguardInterfaceNameDefault)
+					return out
+				}, "10s", "100ms").Should(BeEmpty())
+			}
+
+			// Change WireGuard configuration.
+			ifaceName := "wg0"
+			mtu := 1400
+			rule := 100
+			port := 28150
+			ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+			defer cancel()
+			fc, err := client.FelixConfigurations().Get(ctx, "default", options.GetOptions{})
+			Expect(err).NotTo(HaveOccurred())
+			fc.Spec.WireguardInterfaceName = ifaceName
+			fc.Spec.WireguardMTU = &mtu
+			fc.Spec.WireguardListeningPort = &port
+			fc.Spec.WireguardRoutingRulePriority = &rule
+			_, err = client.FelixConfigurations().Update(ctx, fc, options.SetOptions{})
+			Expect(err).NotTo(HaveOccurred())
+
+			enableDisableWireguard(client, true)
+
+			// The Wireguard device should appear with default MTU, etc.
+			for _, felix := range felixes {
+				Eventually(func() string {
+					out, _ := felix.ExecOutput("ip", "-d", "link", "show", ifaceName)
+					return out
+				}, "10s", "100ms").Should(ContainSubstring(fmt.Sprintf("mtu %d", mtu)))
+			}
+
+			// Expect the settings to be changed on the device.
+			for _, felix := range felixes {
+				Eventually(func() string {
+					out, _ := felix.ExecOutput("ip", "-d", "link", "show", wireguardInterfaceNameDefault)
```

Comment:
For situations where you check for zero value, you also need to check for no error.